src/ASM/ASMs2DMatlab.C

```diff
@@ -0,0 +1,211 @@
+  {
+    std::cerr <<" *** ASMs2DMatlab::read: Not a matlab file."<< std::endl;
+    return false;
+  }
+
+  std::getline(is,cline);
+  while (readLine(is,cline,'['))
+    if (cline.find("Node=") == cline.length()-5)
+    {
+      while (readLine(is,cline))
+      {
+        size_t id; Vec3 X;
+        std::stringstream(cline) >> id >> X;
+#if SP_DEBUG > 1
+        std::cout << id <<": "<< X << std::endl;
+#endif
+	this->setCoord(id,X);
+        if (cline.find_last_of(';') == cline.length()-1) break;
+      }
+#ifdef SP_DEBUG
+      std::cout <<"Read "<< nnod <<" nodes."<< std::endl;
+#endif
+    }
+    else if (cline.find("Element=") == cline.length()-8)
+    {
+      while (readLine(is,cline))
+      {
+        size_t id; int node;
+        IntVec mnpc; mnpc.reserve(4);
+        std::istringstream selem(cline);
+        selem >> id >> node;
+        while (selem)
+        {
+          mnpc.push_back(node-1);
+          selem >> node;
+        }
+#if SP_DEBUG > 1
+        std::cout << id <<":";
+        for (int n : mnpc) std::cout <<" "<< n;
+	std::cout << std::endl;
+#endif
+        if (mnpc.size() == 4)
+          std::swap(mnpc[2],mnpc[3]);
+        else
+        {
+          std::cerr <<"  ** ASMs2DMatlab::read: "<< mnpc.size()
+                    <<"-noded elements not supported (ignored).\n";
+          continue;
+        }
+        if (id > myMNPC.size()) myMNPC.resize(id);
+        myMNPC[id-1] = mnpc;
+	if (cline.find_last_of(';') == cline.length()-1) break;
+      }
+      nel = myMNPC.size();
+#ifdef SP_DEBUG
+      std::cout <<"Read "<< nel <<" elements."<< std::endl;
+#endif
+    }
+    else if (cline.find_first_of('=') == cline.length()-1)
+    {
+      std::string setname = cline.substr(0,cline.length()-1);
+      if (readLine(is,cline,']'))
+      {
+	size_t i = 0; // Remove the '...' continuation markers
+	while ((i = cline.find_first_of('.',i)) != std::string::npos)
+	  cline.erase(i,1);
+
+	int node;
+	IntVec& nodes = nodeSets[setname];
+        std::istringstream selem(cline);
+        selem >> node;
+        while (selem)
+        {
+          nodes.push_back(node);
+          selem >> node;
+        }
+#if SP_DEBUG > 1
+	std::cout <<"Node set \""<< setname <<"\":";
+        for (int n : nodes) std::cout <<" "<< n;
+	std::cout << std::endl;
+#endif
+        std::getline(is,cline);
+      }
+    }
+
+  return true;
+}
+
+
+bool ASMs2DMatlab::generateFEMTopology ()
+{
+  p1 = p2 = 2; // So far only linear elements supported
+
+  nx = sqrt(nnod);
+  ny = nnod / nx;
+
+  myMLGN.resize(nnod);
+  myMLGE.resize(nel);
+
+  std::iota(myMLGN.begin(),myMLGN.end(),gNod+1);
+  std::iota(myMLGE.begin(),myMLGE.end(),gEl+1);
+
+  gNod += nnod;
+  gEl  += nel;
+
+  return true;
+}
+
+
+int ASMs2DMatlab::getNodeSetIdx (const std::string& setName) const
+{
+  std::map<std::string,IntVec>::const_iterator it = nodeSets.begin();
+  for (int i = 1; it != nodeSets.end(); i++, ++it)
+    if (it->first == setName) return i;
```

Comment:
True, but the assumption is that the nodeSets is defined only by the read method, and after that left untouched. Therefore the access method only returns a const reference to it.